src/_zkapauthorizer/replicate.py

```diff
@@ -23,14 +23,77 @@
 ]
 
 from collections.abc import Awaitable
+from io import BytesIO
 from typing import BinaryIO, Callable
 
+import cbor2
+from attrs import frozen
 from twisted.python.lockfile import FilesystemLock
 
 from .config import REPLICA_RWCAP_BASENAME
 from .tahoe import Tahoe, attenuate_writecap
 
 
+@frozen
+class Change:
+    """
+    Represent an item in a replication event stream
+    """
+
+    sequence: int  # the sequence-number of this event
+    statement: str  # the SQL statement string
+
+
+@frozen
+class EventStream:
+    """
+    A series of database operations represented as `Change` instances.
+    """
+
+    changes: tuple[Change]
+
+    def highest_sequence(self):
+        """
+        :return int: the highest sequence number in this EventStream (or
+            None if there are no events)
+        """
+        if not self.changes:
+            return None
+        return max(change.sequence for change in self.changes)
+
+    # XXX or should we pass in a writable stream to use instead?
```

Comment:
I think the only way we could actually realize a benefit is if this could stream right into treq ... so, an `IProducer`?  which I think is still kind of tedious and fiddly.  So probably an optimization for later.

Reply:
`treq` can take file-likes too, so anything from the `io.*` family of ... interfaces? base-classes? was more along the lines I was thinking (although I didn't go far enough to actually-try). But, yeah, could change the API later, it's just internal.